File: src/entities/course/core/rules.ts
import {
  InvalidCourseColorError,
  InvalidCourseDescriptionError,
  InvalidCourseIdError,
  InvalidCourseNameError,
  InvalidCourseStatusError,
  InvalidCourseTimelineError,
  InvalidCourseTopicError,
} from './exceptions'
import type { CourseStatus } from './schema'

export const COURSE_STATUSES: CourseStatus[] = ['draft', 'in_progress', 'completed']
export const DEFAULT_COURSE_STATUS: CourseStatus = 'draft'
export const MAX_TOPIC_LENGTH = 80

export function validateCourseName(name: string): string {
  const value = name.trim()
  if (value.length < 3 || value.length > 120)
    throw new InvalidCourseNameError('Название курса должно содержать от 3 до 120 символов')
  return value
}
export function validateCourseDescription(description: string | null): string | null {
  if (description === null) return null
  const value = description.trim()
  if (value.length > 2000)
    throw new InvalidCourseDescriptionError('Описание курса не должно превышать 2000 символов')
  return value || null
}
export function validateCourseId(id: string): string {
  const value = id.trim()
  if (!value) throw new InvalidCourseIdError('Идентификатор курса не может быть пустым')
  return value
}
export function validateCourseTimeline(createdAt: number, updatedAt: number): void {
  if (updatedAt < createdAt)
    throw new InvalidCourseTimelineError('Дата обновления не может быть раньше даты создания')
}
export function validateCourseTopic(topic: string | null): string | null {
  if (topic === null) return null
  const value = topic.trim()
  if (!value) return null
  if (value.length > MAX_TOPIC_LENGTH)
    throw new InvalidCourseTopicError(`Тема курса не должна превышать ${MAX_TOPIC_LENGTH} символов`)
  return value
}
export function validateCourseColor(color: string | null): string | null {
  if (color === null) return null
  const value = color.trim()
  if (!value) return null
  if (!/^#[0-9a-fA-F]{6}$/.test(value))
    throw new InvalidCourseColorError('Цвет курса должен быть в формате #RRGGBB')
  return value.toLowerCase()
}
export function validateCourseStatus(status: CourseStatus | string): CourseStatus {
  const value = status.trim().toLowerCase()
  if (!COURSE_STATUSES.includes(value as CourseStatus))
    throw new InvalidCourseStatusError(
      `Статус курса должен быть одним из: ${COURSE_STATUSES.join(', ')}`,
    )
  return value as CourseStatus
}

File: src-tauri/src/services/course/rules.rs
// Доменные правила и валидаторы для Course.
// Аналог CourseRules из platform/lib/entities/course/rules.dart.

use super::exceptions::{
    InvalidCourseColorException, InvalidCourseDescriptionException,
    InvalidCourseIdentifierException, InvalidCourseNameException, InvalidCourseStatusException,
    InvalidCourseTimelineException, InvalidCourseTopicException,
};

/// Статус по умолчанию для нового курса.
pub const DEFAULT_COURSE_STATUS: &str = "draft";
/// Допустимые значения статуса курса.
pub const COURSE_STATUSES: [&str; 3] = ["draft", "in_progress", "completed"];

pub struct CourseRules;

impl CourseRules {
    pub const MIN_NAME_LENGTH: usize = 3;
    pub const MAX_NAME_LENGTH: usize = 120;
    pub const MAX_DESCRIPTION_LENGTH: usize = 2000;
    pub const MAX_TOPIC_LENGTH: usize = 80;

    /// Проверяет и нормализует имя курса.
    pub fn validate_course_name(name: &str) -> Result<String, InvalidCourseNameException> {
        let normalized = name.trim().to_string();
        if normalized.is_empty() {
            return Err(InvalidCourseNameException {
                message: "Course name must not be empty.".into(),
            });
        }
        if normalized.len() < Self::MIN_NAME_LENGTH {
            return Err(InvalidCourseNameException {
                message: format!(
                    "Course name must be at least {} characters long.",
                    Self::MIN_NAME_LENGTH
                ),
            });
        }
        if normalized.len() > Self::MAX_NAME_LENGTH {
            return Err(InvalidCourseNameException {
                message: format!(
                    "Course name must not exceed {} characters.",
                    Self::MAX_NAME_LENGTH
                ),
            });
        }
        Ok(normalized)
    }

    /// Проверяет и нормализует описание курса.
    pub fn validate_course_description(
        description: Option<&str>,
    ) -> Result<Option<String>, InvalidCourseDescriptionException> {
        match description {
            None => Ok(None),
            Some(d) => {
                let normalized = d.trim().to_string();
                if normalized.is_empty() {
                    return Ok(None);
                }
                if normalized.len() > Self::MAX_DESCRIPTION_LENGTH {
                    return Err(InvalidCourseDescriptionException {
                        message: format!(
                            "Course description must not exceed {} characters.",
                            Self::MAX_DESCRIPTION_LENGTH
                        ),
                    });
                }
                Ok(Some(normalized))
            }
        }
    }

    /// Проверяет и нормализует идентификатор курса.
    pub fn validate_course_id(id: &str) -> Result<String, InvalidCourseIdentifierException> {
        let normalized = id.trim().to_string();
        if normalized.is_empty() {
            return Err(InvalidCourseIdentifierException {
                message: "Course id must not be empty.".into(),
            });
        }
        Ok(normalized)
    }

    /// Проверяет корректность временных меток курса.
    pub fn validate_course_timeline(
        created_at: i64,
        updated_at: i64,
    ) -> Result<(), InvalidCourseTimelineException> {
        if updated_at < created_at {
            return Err(InvalidCourseTimelineException {
                message: "updated_at cannot be earlier than created_at.".into(),
            });
        }
        Ok(())
    }

    /// Проверяет и нормализует тему курса (пустая/пробельная → None).
    pub fn validate_course_topic(
        topic: Option<&str>,
    ) -> Result<Option<String>, InvalidCourseTopicException> {
        match topic {
            None => Ok(None),
            Some(t) => {
                let normalized = t.trim().to_string();
                if normalized.is_empty() {
                    return Ok(None);
                }
                if normalized.len() > Self::MAX_TOPIC_LENGTH {
                    return Err(InvalidCourseTopicException {
                        message: format!(
                            "Course topic must not exceed {} characters.",
                            Self::MAX_TOPIC_LENGTH
                        ),
                    });
                }
                Ok(Some(normalized))
            }
        }
    }

    /// Проверяет и нормализует цвет карточки (hex-формат `#RRGGBB`).
    pub fn validate_course_color(
        color: Option<&str>,
    ) -> Result<Option<String>, InvalidCourseColorException> {
        const HEX_DIGITS: &str = "0123456789abcdefABCDEF";
        match color {
            None => Ok(None),
            Some(c) => {
                let normalized = c.trim().to_string();
                if normalized.is_empty() {
                    return Ok(None);
                }
                let valid = normalized.len() == 7
                    && normalized.starts_with('#')
                    && normalized[1..].chars().all(|ch| HEX_DIGITS.contains(ch));
                if !valid {
                    return Err(InvalidCourseColorException {
                        message: "Course color must be in #RRGGBB format.".into(),
                    });
                }
                Ok(Some(normalized.to_lowercase()))
            }
        }
    }

    /// Проверяет статус курса (draft | in_progress | completed).
    pub fn validate_course_status(status: &str) -> Result<String, InvalidCourseStatusException> {
        let normalized = status.trim().to_lowercase();
        if COURSE_STATUSES.contains(&normalized.as_str()) {
            Ok(normalized)
        } else {
            Err(InvalidCourseStatusException {
                message: format!(
                    "Course status must be one of: {}.",
                    COURSE_STATUSES.join(", ")
                ),
            })
        }
    }
}
